mlapi.py

```diff
@@ -0,0 +1,100 @@
+import numpy as np
+import pandas as pd
+import scipy
+import joblib
+from sklearn.preprocessing import RobustScaler
+from flask import Flask, request, jsonify
+
+mlapi=Flask(__name__)
+
+@mlapi.route("/get_question/<int:cnt>",methods=["GET"])
+def recommend_question(cnt):
+    og_dat1=pd.read_csv(r"og_dataset.csv")
+    prob_dat=pd.DataFrame(scipy.sparse.load_npz(r"prob_table.npz").toarray(),columns=pd.read_csv(r"prob_table_col_labels.csv",sep=',')['0'])
+    prob_dat=prob_dat.sort_values(['Probability']).reset_index(drop=True)
+    recomm=og_dat1[og_dat1['Cluster_label']==prob_dat.iloc[cnt,0]][['title','difficulty','related_topics']].to_json(orient='records')
+    return jsonify(recomm), 200
+
+@mlapi.route("/retrain")
+def model_update(): # process time,update the clusters and update prob_table
+    #loading tables
+    og_dat=pd.read_csv(r"og_dataset.csv")
+    user_time_label=pd.read_csv(r"user_time_taken_col_labels.csv",sep=',')
+    time_dat=pd.DataFrame(scipy.sparse.load_npz(r"user_time_taken.npz").toarray(),columns=user_time_label['0'])
+    user_prob_label=pd.read_csv(r"user_prob_table_col_labels.csv",sep=',')
+    uprob_table=pd.DataFrame(scipy.sparse.load_npz(r"user_prob_table.npz").toarray(),columns=user_prob_label['0'])
+
+    #processing time_taken feature
+
+    #time_dat.loc[:,'time_taken']=time_dat.iloc[:,2:].mean(axis=1)
+    og_dat.loc[:,'time_taken']=time_dat.iloc[:,2:].mean(axis=1)
+    #time_dat=time_dat.loc[:,('id','Cluster_label','time_taken')]
+    #user_time_label=user_time_label.iloc[:3,0]
+
+    #retrain the model
+    scaled_data=tran_mod.fit_transform(og_dat)
+    mod.fit(scaled_data)
+
+    #update the probability table
+    og_dat['Cluster_label']=mod.labels_
+    uprob_table['Cluster_label']=mod.labels_
+    prob_table=uprob_table.iloc[:,1:].groupby(['Cluster_label']).sum()
+    prob_table=prob_table.divide(uprob_table.iloc[:,:2].groupby(['Cluster_label']).count().loc[:,"id"],axis=0) #check
+    prob_table=prob_table.mean(axis=1).round(5).abs().reset_index()
+    prob_table.rename(columns={0:'Probability'},inplace=True)
+
+    #saving all tables
+    spar=scipy.sparse.csc_matrix(prob_table)
+    scipy.sparse.save_npz(r"prob_table.npz",spar,True)
+    spar2=scipy.sparse.csc_matrix(time_dat)
+    scipy.sparse.save_npz(r"user_time_taken.npz",spar2,True)
+    spar3=scipy.sparse.csc_matrix(uprob_table)
+    scipy.sparse.save_npz(r"user_prob_table.npz",spar3,True)
+    og_dat.to_csv(r"og_dataset.csv",sep=',')
+    user_time_label.to_csv(r"user_time_taken_col_labels.csv",sep=',')
+    user_prob_label.to_csv(r"user_prob_table_col_labels.csv",sep=',')
+
+    return jsonify({"retrain_model":"success","message":"Model has been updated"})
+
+@mlapi.route("/data_process_task/",methods=['POST'])
+def table_update(): #replace time of og_dat after processing and add 'like' to og data
+    data=request.get_json()
```

Comment:
Whenever there is a json input involved, please try to use `dict.get("key", "default")` to return the default when that field is omitted.